Modules/posixmodule.c

```diff
@@ -4800,43 +4802,70 @@ os__path_isdir_impl(PyObject *module, PyObject *path)
     }
 
     Py_BEGIN_ALLOW_THREADS
-    if (_path.fd != -1) {
-        hfile = _Py_get_osfhandle_noraise(_path.fd);
-        close_file = FALSE;
-    }
-    else {
-        hfile = CreateFileW(_path.wide, FILE_READ_ATTRIBUTES, 0, NULL,
-                            OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+    if (_path.wide) {    
+        if (_Py_GetFileInformationByName(_path.wide, FileStatBasicByNameInfo,
+                                         &statInfo, sizeof(statInfo))) {
+            if (!(statInfo.FileAttributes & FILE_ATTRIBUTE_REPARSE_POINT)) {
+                slow_path = FALSE;
+                result = statInfo.FileAttributes & FILE_ATTRIBUTE_DIRECTORY;
+            } else if (!(statInfo.FileAttributes & FILE_ATTRIBUTE_DIRECTORY)) {
+                slow_path = FALSE;
+                result = 0;
+            }
+        } else {
+            switch(GetLastError()) {
+                case ERROR_FILE_NOT_FOUND:
+                case ERROR_PATH_NOT_FOUND:
+                case ERROR_NOT_READY:
+                case ERROR_BAD_NET_NAME:
+                    /* These errors aren't worth retrying with the slow path */
+                    slow_path = FALSE;
+                    result = 0;
+                case ERROR_NOT_SUPPORTED:
+                    /* indicates the API couldn't be loaded */
+                    break;
```

Comment:
Why is this case highlighted?

Note that `ERROR_NOT_SUPPORTED` could also be returned if the device doesn't support the by-name query. Though in that case the error code is more likely to be `ERROR_INVALID_FUNCTION` or `ERROR_INVALID_PARAMETER`.

Reply:
Ah, I did not realize that. Thanks! 

I highlighted this case since it was also [highlighted this way](https://github.com/python/cpython/blob/main/Modules/posixmodule.c#L2056) were `Py_GetFileInformationByName` was called in the win32 xstat implementation earlier in the file. Would it be better to alter this comment to say something like "indicates the API couldn't be loaded or the device doesn't support the by-name query" or would it be better just to remove the comment entirely?

Reply:
Honestly, it's probably better to pick a different error code for when the API is not supported. We don't _really_ care, but it's worth recognising when it happened.

Maybe let's return `ERROR_NOT_SUPPORTED | (1 << 29)` (see [SetLastError docs](https://learn.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-setlasterror) for the bit 29 reference). Then just leave out the cases that capture and ignore it.